File: src/runtime/cs/std/os_path_native.cs
using System.IO;
using SysPath = System.IO.Path;

namespace Pytra.CsModule
{
    // Generated std/os_path.cs delegates host bindings through this native seam.
    public static class os_path_native
    {
        public static string join(string a, string b)
        {
            return SysPath.Combine(a, b).Replace('\\', '/');
        }

        public static string dirname(string p)
        {
            return (SysPath.GetDirectoryName(p) ?? "").Replace('\\', '/');
        }

        public static string basename(string p)
        {
            return SysPath.GetFileName(p);
        }

        public static (string, string) splitext(string p)
        {
            string ext = SysPath.GetExtension(p);
            string stem = p.Substring(0, p.Length - ext.Length);
            return (stem, ext);
        }

        public static string abspath(string p)
        {
            return SysPath.GetFullPath(p).Replace('\\', '/');
        }

        public static bool exists(string p)
        {
            return File.Exists(p) || Directory.Exists(p);
        }
    }
}

File: src/runtime/nim/std/os_path_native.nim
# Native implementation of pytra.std.os_path for Nim.
#
# source: src/pytra/std/os_path.py

import std/os
import std/strutils

proc join*(a: string, b: string): string = joinPath(a, b)

proc dirname*(p: string): string = parentDir(p)

proc basename*(p: string): string = extractFilename(p)

proc splitext*(p: string): (string, string) =
  let (dir, name, ext) = splitFile(p)
  result = (joinPath(dir, name), ext)

proc abspath*(p: string): string = absolutePath(p)

proc exists*(p: string): bool = fileExists(p) or dirExists(p)
